compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -12845,6 +12845,148 @@ TR::Node *su2dSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s)
    }
 
 
+bool canTransformOp(bool canTransformAdd, bool canTransformSub, TR::Node *node, TR::Simplifier *s)
+   {
+   if (!(canTransformAdd || canTransformSub))
+      {
+      if (s->trace())
+         traceMsg(s->comp(),
+                  "\nEliminating add/sub under order compare node n%dn failed due to overflow\n",
+                  node->getGlobalIndex());
+
+      return false;
+      }
+   return true;
+   }
+
+TR::Node *
+arithUnderSignedAndUnsignedCompareHandler(TR::Node *opNode, TR::Node *secondChild, TR::Simplifier *s, TR::Node *node,
+                                          bool isSigned, bool isAddOp, bool isSubOp)
+   {
+
+   TR::DataType DataType = secondChild->getDataType();
+   if (!isSigned)
+      {
+      uint64_t oldUConst1 = 0, oldUConst2 = 0, unsignedMax = 0;
+      switch (DataType)
+         {
+         case TR::Int8:
+            oldUConst1 = opNode->getSecondChild()->getUnsignedByte();
+            oldUConst2 = secondChild->getUnsignedByte();
+            unsignedMax = TR::getMaxUnsigned<TR::Int8>();
+            break;
+         case TR::Int16:
+            oldUConst1 = opNode->getSecondChild()->getUnsignedShortInt();
+            oldUConst2 = secondChild->getUnsignedShortInt();
+            unsignedMax = TR::getMaxUnsigned<TR::Int16>();
+            break;
+         case TR::Int32:
+            oldUConst1 = opNode->getSecondChild()->getUnsignedInt();
+            oldUConst2 = secondChild->getUnsignedInt();
+            unsignedMax = TR::getMaxUnsigned<TR::Int32>();
+            break;
+         case TR::Int64:
+            oldUConst1 = opNode->getSecondChild()->getUnsignedLongInt();
+            oldUConst2 = secondChild->getUnsignedLongInt();
+            unsignedMax = TR::getMaxUnsigned<TR::Int64>();
+            break;
+         default:
+            if (s->trace())
+               traceMsg(s->comp(),
+                        "\nEliminating add/sub under compare node n%dn failed due to opcode data type\n",
+                        node->getGlobalIndex());
+
+            TR_ASSERT_FATAL(false, "\nEliminating add/sub under compare node n%dn failed due to opcode data type\n",
+                            node->getGlobalIndex());
+            break;
+         }
+      //overflow handling:
+      //Must detect overflow without causing overflow
+      //Assuming oldUConst1+oldUConst2>unsignedMax --> resulting in overflow,
+      //We can identify such overflow from
+      //oldUConst2>unsignedMax-oldUConst1 without causing overflow in the calculation itself
+      bool canTransformAdd = isAddOp
```

Comment:
This doesn't look quite right... just because a < b does not mean a+b < unsignedMax...

Reply:
to my understanding
preopt:
x+a<b
postopt:
x<b-a
in unsigned, a and b must be greater than 0, and therefore underflow occurs, when a is greater than b
or 
!(oldUConst2<oldUConst1)

Reply:
So for the undeflow case I agree. eg that b > a means a-b will not underflow doing unsigned math. Please update the comments. Also make clear the behaviour of x+a / x-a is covered by the cannot overflow flag so we are only concerned about underflow / overflow when transforming the constants...

Reply:
I think I clearly updated the comments. my only remaining concern is we handle a cannotOverflow flag but I can't find an equivalent cannotUnderflow flag. is there a reason for this?

Reply:
@andrewcraik what is the conclusion on this comment and the following one?

https://github.com/eclipse/omr/pull/4305#discussion_r325735442